script/update-content-versions-and-check-for-archived-slugs.ts

```diff
@@ -0,0 +1,112 @@
+import { ARCHIVE_URL, REACT_APP_ARCHIVE_URL, REACT_APP_OS_WEB_API_URL } from '../src/config';
+import books from '../src/config.books';
+import createArchiveLoader from '../src/gateways/createArchiveLoader';
+import createOSWebLoader from '../src/gateways/createOSWebLoader';
+import { Redirects } from '../src/redirects/types';
+
+const { bookId, newVersion } = argv as any as {
+  bookId: string
+  newVersion: string | number;
+};
+
+const booksPath = path.resolve(__dirname, '../src/config.books.json');
+const redirectsPath = path.resolve(__dirname, '../src/redirects/');
+
+const bookLoader = makeUnifiedBookLoader(
+  createArchiveLoader(`${ARCHIVE_URL}${REACT_APP_ARCHIVE_URL}`),
+  createOSWebLoader(`${ARCHIVE_URL}${REACT_APP_OS_WEB_API_URL}`)
+);
+
+async function updateRedirections(_bookId: string, currentVersion: string, _newVersion: string) {
+  if (currentVersion === _newVersion) {
+    return 0;
+  }
+
+  const { tree: currentTree, slug: bookSlug } = await bookLoader(_bookId, currentVersion)
+    .catch((error) => {
+      // tslint:disable-next-line: no-console
+      console.log(`error while loading book ${_bookId} with defaultVersion ${currentVersion}`);
+      throw error;
+    });
+
+  const { tree: newTree } = await bookLoader(_bookId, _newVersion)
+    .catch((error) => {
+      // tslint:disable-next-line: no-console
+      console.log(`error while loading book ${_bookId} with newVersion ${_newVersion}`);
+      throw error;
+    });
+
+  const redirectsBookPath = path.resolve(redirectsPath, bookId + '.json');
+
+  try {
+    // Create file with empty array only if it does not exixts
+    fs.writeFileSync(redirectsBookPath, '[]', { encoding: 'utf8', flag: 'wx' });
+  } catch {
+    // This will throw if file exists
+  }
+
+  const redirects: Redirects = require(redirectsBookPath);
+
+  const flatCurrentTree = flattenArchiveTree(currentTree);
+
+  const findRedirect = (section: LinkedArchiveTreeNode) => (
+    { pageId, bookId: pageBookId, pathname }: Redirects[0]
+  ) => pageId === section.id && pageBookId === _bookId && pathname.split('/').pop() === section.slug;
+
+  const formatSection = (section: LinkedArchiveTreeNode) => ({
+    bookId: _bookId,
+    pageId: section.id,
+    pathname: content.getUrl({ book: { slug: bookSlug }, page: { slug: section.slug } }),
+  });
+
+  let countNewRedirections = 0;
+  for (const section of flatCurrentTree) {
+    const { slug } = findArchiveTreeNodeById(newTree, section.id) || {};
+    if (
+      (slug && slug !== section.slug)
+      && !redirects.find(findRedirect(section))
+    ) {
+      redirects.push(formatSection(section));
+      countNewRedirections++;
+    }
+  }
+
+  fs.writeFileSync(redirectsBookPath, JSON.stringify(redirects, undefined, 2) + '\n', 'utf8');
+
+  return countNewRedirections;
+}
+
+async function processBook() {
+  const { defaultVersion } = books[bookId] || {};
+
+  if (defaultVersion === newVersion.toString()) {
```

Comment:
you're also doing this check above, do you need to do it twice?

Reply:
I wanted to separate logic for updating redirects and updating the config file
I agree that currently, this doesn't matter, but I like when functions don't assume anything unless they are written inside of another function